### PR TITLE
 Support multiple Mesos master URLs. 

### PR DESCRIPTION
**Pull request by jeschkies**
Summary:
This change makes the Mesos master URLs a list of addresses to support HA mode.
All URLs are tried one after another. It also makes the `MesosClientSettings` a simple
class to ensure backwards compatibility.

JIRA issues: DCOS-53671

### CODE REVIEW

File: examples/core-hello-world/src/main/resources/application.conf [hunk 1]
```diff
@@ -1,8 +1,8 @@
 # This is still the default configuration but it's good to see what can be changed
 mesos-client {
 
-  master-url: "127.0.0.1:5050"
-  master-url: ${?MESOS_MASTER_URL}
+  master-url: [ "127.0.0.1:5050" ]
```
Comment (timcharper) [inline examples/core-hello-world/src/main/resources/application.conf]: Let's put a protocol here (and force a protocol). Maybe another PR?

Comment (timcharper) [reply examples/core-hello-world/src/main/resources/application.conf]: Perhaps also we should change the key name from `master-url` to `master-urls` ?

File: mesos-client/src/main/scala/com/mesosphere/mesos/conf/MesosClientSettings.scala [hunk 2]
```diff
@@ -1,25 +1,57 @@
...
+    val sourceBufferSize: Int) {
+
+  private def copy(
+      masters: Seq[URL] = this.masters,
+      redirectRetries: Int = this.redirectRetries,
+      idleTimeout: FiniteDuration = this.idleTimeout,
+      sourceBufferSize: Int = this.sourceBufferSize) =
+    new MesosClientSettings(masters, redirectRetries, idleTimeout, sourceBufferSize)
+
+  def withMasters(urls: URL*): MesosClientSettings = copy(masters = urls)
```
Comment (timcharper) [inline mesos-client/src/main/scala/com/mesosphere/mesos/conf/MesosClientSettings.scala]: Let's please just have one method. `withMaster` is unclear whether it appends or replaces the current masters.

File: examples/core-hello-world/src/main/resources/application.conf (hunk 1 above)
Comment (jeschkies) [reply examples/core-hello-world/src/main/resources/application.conf]: That's a bug from my side. Thanks.

File: mesos-client/src/main/scala/com/mesosphere/mesos/conf/MesosClientSettings.scala (hunk 2 above)
Comment (jeschkies) [reply mesos-client/src/main/scala/com/mesosphere/mesos/conf/MesosClientSettings.scala]: Java does not support var args. `withMasters(urls: URL*)` is a `withMasters(scala.collection.Seq<URL> urls)` in Java which is annoying to use.

File: examples/core-hello-world/src/main/resources/application.conf [hunk 3]
```diff
@@ -1,10 +1,10 @@
 # This is still the default configuration but it's good to see what can be changed
 mesos-client {
 
-  master-url: "127.0.0.1:5050"
-  master-url: ${?MESOS_MASTER_URL}
+  master-url: [ "http://127.0.0.1:5050" ]
+  master-url: [ ${?MESOS_MASTER_URL} ]
```
Comment (zen-dog) [inline examples/core-hello-world/src/main/resources/application.conf]: Does env var override work for arrays?

Comment (zen-dog) [reply examples/core-hello-world/src/main/resources/application.conf]: To answer my own question - it does, but you can only pass **one** value.

**Review by zen-dog — APPROVED** [review]
If it's green - go for it! 🚢

File: examples/core-hello-world/src/main/resources/application.conf (hunk 3 above)
Comment (timcharper) [reply examples/core-hello-world/src/main/resources/application.conf]: Oof, yeah. A quick search reveals that we probably need to resort to splitting for env-var overrides :'(

https://github.com/lightbend/config/issues/163

**Review by timcharper — CHANGES_REQUESTED** [review]
Lets get clarity on how we will solve the environment variable overrides; I'm wondering if we will want to do the comma delimited authority approach. I've a thread open in the Mesos slack: https://mesos.slack.com/archives/C1K2TAJ1L/p1557739907081100

File: mesos-client/src/main/resources/reference.conf [hunk 4]
```diff
@@ -1,10 +1,10 @@
 mesos-client {
 
-  # Mesos master host:port
-  master-url: "127.0.0.1:5050"
+  # Comma separated list of Mesos master protocol://host:port,...
```
Comment (timcharper) [inline mesos-client/src/main/resources/reference.conf]: docs doesn't seem to match the implementation

**Review by timcharper — CHANGES_REQUESTED** [review]

File: mesos-client/src/main/resources/reference.conf (hunk 4 above)
Comment (timcharper) [inline mesos-client/src/main/resources/reference.conf]:
```suggestion
  # Comma separated list of Mesos master protocol://host1:port1,protocol://host2:port2,...
  # protocol must either be http or https
```

**Review by timcharper — APPROVED** [review]

**Pull request merged**